Remove `glom` usage from `FormioData`

viktorvanwijk opened the pull request:
Closes #5179

**Changes**

Remove usage of `glom` from `__getitem__` and `__setitem__` in `FormioData`

**Checklist**

Check off the items that are completed or not relevant.

- Impact on features

  - [x] Checked copying a form
  - [x] Checked import/export of a form
  - [x] Config checks in the configuration overview admin page
  - [x] Problem detection in the admin email digest is handled

- Release management

  - [x] I have labelled the PR as "needs-backport" accordingly

- I have updated the translations assets (you do NOT need to provide translations)

  - [x] Ran `./bin/makemessages_js.sh`
  - [x] Ran `./bin/compilemessages_js.sh`

- Dockerfile/scripts

  - [x] Updated the Dockerfile with the necessary scripts from the `./bin` folder

- Commit hygiene

  - [x] Commit messages refer to the relevant Github issue
  - [x] Commit messages explain the "why" of change, not the how

viktorvanwijk commented:
# Performance test result

Performance is tested using `FormioDataTests.test_performance`. Below are the results per method.

Result interpretation:
- Current: the current implementation (`FormioData`)
- Full glom: the current implementation without the ‘bandaid’ fixes (`FormioDataWithoutGlom`)
- Custom: the new implementation, presented by the code changes shown here (`FormioDataCustom`)
- Control: regular dict lookup with flat structure (`{“foo.bar”: “baz”}`)

### `__getitem__`
```python
    def __getitem__(self, key: str):
        # TODO-5179: do we want to check the key here as well?
        # self._check_key(key)

        value = self.data
        for k in key.split("."):
            if not isinstance(value, dict):
                raise KeyError(f"Key '{key}' is not present in the data")
            try:
                value = value[k]
            except KeyError:
                raise KeyError(f"Key '{key}' is not present in the data")

        return value
```
![image](https://github.com/user-attachments/assets/5d76ff0d-87bf-4fdc-b9af-9f0859407438)

Comparing 'Current' to 'Custom', we can see a significant increase in performance for nested keys. Note that if we want the non-nested key lookup to be as fast as the current implementation, we need to keep the "bandaid fix":
```python
if "." not in key:
    return self.data[key]
```

### `__setitem__` (for loop)
```python
```

sergei-maertens reviewed (changes requested):
Thanks for the thorough analysis, some thoughts:

**Hot code path**

Arguably the non-nested key access (`"foo"` rather than `"foo.bar"`) is the most used in forms, so I'd opt for keeping the "bandaid" fix as an extra optimization. From your benchmarks, not doing it makes the code roughly twice as slow (+100% execution time), and that may add up in large forms. The maintenance overhead of this extra check is minimal and properly contained in the data structure, so I'd reckon this is a good tradeoff.

**Key type checking**

There's now an extra method call `self._check_key(key)` which causes some overhead and slowdown, and I think it mostly just provides developer type safety. I haven't seen any crashes in Sentry. So, we have the `isinstance` check + the method call overhead, and I think we can avoid it:

* short term: change it to an inline `assert isinstance(key, str)` - this enforces it at runtime during development and will be caught in the test suite if people do silly stuff, but allows us to use the [uwsgi optimize](https://uwsgi-docs.readthedocs.io/en/latest/Options.html#optimize) option to strip out those `asserts` in production (same as using the `python -O` option).

* long term: type checking with pyright! we have the type annotation for the key already, and if we get all of our code accessing FormioData properly type checked, this will also be caught before it causes problems and the beauty is that it has zero runtime impact.

**Recursion vs for-loop**

Agree with your conclusion, and it's more obvious if we do the hot-code path optimization.

**Missing cases**

I think one case that we are missing that `glom` covers is index access, e.g. `foo.0.bar`. In Formio itself, this actually still leads to nested dictionaries, but those access paths are ultimately what's driving data access in edit grids - a repeating group with key `foo` has items (accessed by index) which themselves are objects/dicts with keys (`bar). Though this is probably not relevant - I can only imagine some JSON logic triggers with paths to a particular item (which in itself is kind of weird), and that has its own data resolution mechanism separate from `FormioData`.

viktorvanwijk commented on src/openforms/formio/datastructures.py:
```diff
@@ -157,7 +157,7 @@ class FormioData(UserDict):
+    details.
     """
 
     data: dict[str, JSONValue]
```
Yeah it should be extended to include the Python types :)

viktorvanwijk commented:
> Arguably the non-nested key access ("foo" rather than "foo.bar") is the most used in forms, so I'd opt for keeping the "bandaid" fix as an extra optimization. From your benchmarks, not doing it makes the code roughly twice as slow (+100% execution time), and that may add up in large forms. The maintenance overhead of this extra check is minimal and properly contained in the data structure, so I'd reckon this is a good tradeoff.

Yeah, fair enough. I was doubting a little if we wanted to get rid of this "bandaid", so I left it out for now. But the performance gain is significant :zap:

> short term: change it to an inline assert isinstance(key, str) - this enforces it at runtime during development and will be caught in the test suite if people do silly stuff, but allows us to use the [uwsgi optimize](https://uwsgi-docs.readthedocs.io/en/latest/Options.html#optimize) option to strip out those asserts in production (same as using the python -O option).

TIL something :)

> I think one case that we are missing that glom covers is index access, e.g. foo.0.bar. In Formio itself, this actually still leads to nested dictionaries, but those access paths are ultimately what's driving data access in edit grids - a repeating group with key foo has items (accessed by index) which themselves are objects/dicts with keys (bar). Though this is probably not relevant - I can only imagine some JSON logic triggers with paths to a particular item (which in itself is kind of weird), and that has its own data resolution mechanism separate from FormioData

Yeah I wasn't sure if this was necessary at the moment, so I left it out. While working on #5221, however, I have found a place where this is used: `ComponentNode.value` and `ComponentNode.is_visible`. They both currently still have raw `glom` usage. Would be nice if we can incorporate this inside `FormioData` then :)

viktorvanwijk commented:
# Performance results after feedback
![image](https://github.com/user-attachments/assets/8a4edec2-5688-48ea-ab35-0823b68f878f)


![image](https://github.com/user-attachments/assets/361c1a76-1052-4ec0-9191-2b3839868047)

NOTE: during testing, I have noticed variations of up to 0.01 µs when running the same test multiple times, so please take this into consideration when looking at the numbers for non-nested items

viktorvanwijk commented on src/openforms/formio/tests/test_datastructures.py:
```diff
@@ -9,7 +9,7 @@ class FormioDataTests(TestCase):
     def test_mimicks_dict_interface(self):
         formio_data = FormioData({"foo": "bar"})
 
-        self.assertEqual(formio_data, {"foo": "bar"})
```
I have found usages of `FormioData.items` :)

viktorvanwijk commented on src/openforms/formio/tests/test_datastructures.py:
```diff
@@ -9,7 +9,7 @@ class FormioDataTests(TestCase):
         formio_data = FormioData({"foo": "bar"})
 
-        self.assertEqual(formio_data, {"foo": "bar"})
+        self.assertEqual(formio_data.data, {"foo": "bar"})
```
I have found usages of FormioData.items :)

viktorvanwijk commented:
I started to work on a `Path` class, similar to the one that `glom` has, but we *might* be able to get away with just the list indexing in strings, e.g. `'foo.0.bar'`. Can't tell for sure until I've replaced all the `DataMapping` instances with `FormioData`, so I've left it out for now. It can be added later in #5228 if necessary.
```python
class Path:
    def __init__(self, *args) -> None:
        self.values = ()
        for k in args:
            if isinstance(k, str):
                self.values = self.values + tuple(k.split("."))
            elif isinstance(k, int):
                self.values = self.values + (k,)
            elif isinstance(k, Path):
                self.values = self.values + k.values
            else:
                raise NotImplementedError(f"Unsupported type {type(k)}")

    def __getitem__(self, index: int):
        return self.values[index]

    def __iter__(self):
        yield from self.values

    def __str__(self):
        return str(self.values)

    def __add__(self, other):
        if isinstance(other, Path):
            return Path(self.values + other.values)
```
